### PR TITLE
refactor(iroh): remove flume from iroh-cli and iroh

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

refactor(iroh): remove flume from iroh-cli and iroh

Removes most usages of flume from iroh and all usages of flume from iroh-cli.

## Breaking Changes

None

## Notes & open questions

Note: this does not remove all usages because we need the iroh docs purge PR to be merged before that.

## Change checklist

- [ ] Self-review.
- [ ] ~~Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.~~
- [ ] ~~Tests if relevant.~~
- [ ] ~~All breaking changes documented.~~

### CODE REVIEW

File: iroh-blobs/src/util/progress.rs [hunk 1]
```diff
@@ -518,6 +518,93 @@ impl<T: Send + Sync + 'static> ProgressSender for FlumeProgressSender<T> {
...
+        }
+    }
+
+    /// Returns true if `other` sends on the same `flume` channel as `self`.
+    pub fn same_channel(&self, other: &AsyncChannelProgressSender<T>) -> bool {
+        same_channel(&self.sender, &other.sender)
+    }
+}
+
+fn get_as_ptr<T>(value: &T) -> Option<usize> {
```
Comment (dignifiedquire) [inline iroh-blobs/src/util/progress.rs:560]: maybe add some details of why this works and a test, this is assuming a lot of things

Comment (rklaehn) [reply iroh-blobs/src/util/progress.rs:562]: Added a few tests.

File: iroh-blobs/src/util/progress.rs [hunk 2]
```diff
@@ -518,6 +518,98 @@ impl<T: Send + Sync + 'static> ProgressSender for FlumeProgressSender<T> {
     }
 }
 
+/// A progress sender that uses a flume channel.
```
Comment (matheus23) [inline iroh-blobs/src/util/progress.rs:521]:
```suggestion
/// A progress sender that uses an async channel.
```

File: iroh-blobs/src/util/progress.rs [hunk 3]
```diff
@@ -518,6 +518,98 @@ impl<T: Send + Sync + 'static> ProgressSender for FlumeProgressSender<T> {
...
+    fn clone(&self) -> Self {
+        Self {
+            sender: self.sender.clone(),
+            id: self.id.clone(),
+        }
+    }
+}
+
+impl<T> AsyncChannelProgressSender<T> {
+    /// Create a new progress sender from a flume sender.
```
Comment (matheus23) [inline iroh-blobs/src/util/progress.rs:546]:
```suggestion
    /// Create a new progress sender from an async channel sender.
```

File: iroh-blobs/src/util/progress.rs [hunk 4]
```diff
@@ -518,6 +518,98 @@ impl<T: Send + Sync + 'static> ProgressSender for FlumeProgressSender<T> {
...
+    pub fn new(sender: async_channel::Sender<T>) -> Self {
+        Self {
+            sender,
+            id: std::sync::Arc::new(std::sync::atomic::AtomicU64::new(0)),
+        }
+    }
+
+    /// Returns true if `other` sends on the same `flume` channel as `self`.
```
Comment (matheus23) [inline iroh-blobs/src/util/progress.rs:554]:
```suggestion
    /// Returns true if `other` sends on the same `async_channel` channel as `self`.
```

File: iroh-blobs/src/util/progress.rs [hunk 5]
```diff
@@ -628,3 +720,39 @@ impl<W: AsyncSliceWriter + 'static, F: Fn(u64, usize) -> io::Result<()> + 'stati
...
+    fn test_sender_is_ptr() {
+        assert_eq!(
+            std::mem::size_of::<usize>(),
+            std::mem::size_of::<flume::Sender<u8>>()
+        );
+        assert_eq!(
+            std::mem::align_of::<usize>(),
+            std::mem::align_of::<flume::Sender<u8>>()
+        );
+    }
```
Comment (matheus23) [inline iroh-blobs/src/util/progress.rs:757]: Should we update these tests?

```suggestion
    #[test]
    fn test_sender_is_ptr() {
        assert_eq!(
            std::mem::size_of::<usize>(),
            std::mem::size_of::<async_channel::Sender<u8>>()
        );
        assert_eq!(
            std::mem::align_of::<usize>(),
            std::mem::align_of::<async_channel::Sender<u8>>()
        );
    }
```

File: iroh-blobs/src/util/progress.rs [hunk 6]
```diff
@@ -518,6 +518,93 @@ impl<T: Send + Sync + 'static> ProgressSender for FlumeProgressSender<T> {
...
+        same_channel(&self.sender, &other.sender)
+    }
+}
+
+fn get_as_ptr<T>(value: &T) -> Option<usize> {
+    use std::mem;
+    if mem::size_of::<T>() == std::mem::size_of::<usize>()
+        && mem::align_of::<T>() == mem::align_of::<usize>()
+    {
+        // Safe only if size and alignment requirements are met
```
Comment (matheus23) [inline iroh-blobs/src/util/progress.rs:567]:
```suggestion
        // SAFETY: size and alignment requirements are checked and met
```

**Comment by matheus23** [comment]
Is this the last flume purge PR?
Should we add `flume` to `deny.toml`?

**Comment by rklaehn** [comment]
> Is this the last flume purge PR?

I wish. We still got flume in iroh-blobs and quic-rpc... These three were the easy ones...

File: iroh-blobs/src/util/progress.rs (hunk 2 above)
Comment (matheus23) [reply iroh-blobs/src/util/progress.rs:521]: @rklaehn I think you are still missing this and another quick doc update suggestion :v:

**Review by matheus23 — APPROVED** [review]

**Pull request merged**